Json format application logs for kibana

mburke5678 opened the pull request:
https://bugzilla.redhat.com/show_bug.cgi?id=1710864

Steps taken from:
https://richm.github.io/add-fields-to-index-template
https://richm.github.io/how-to-add-fields-to-kibana

mburke5678 commented on install_config/aggregate_logging.adoc:
```diff
@@ -2054,6 +2054,186 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
+}
+----
+<1> Specify the `.operations.*` or a `project.` index pattern.
+
```
@QiaolingTang 
What would the user put as `template:` for the OPS cluster index pattern?

mburke5678 commented on install_config/aggregate_logging.adoc:
```diff
@@ -2054,6 +2054,186 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
+NAME                                       READY     STATUS    RESTARTS   AGE       IP             NODE                         NOMINATED NODE
+logging-es-data-master-5av030lk-1-2x494    2/2       Running   0          38m       154.128.0.80   ip-153-12-8-6.wef.internal   <none>
+----
+
```
@QiaolingTang @richm 
For a project index pattern, which pod to use?

mburke5678 commented on install_config/aggregate_logging.adoc:
```diff
@@ -2054,6 +2054,186 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
++
+The index patterns download to the *_/usr/share/elasticsearch/index_patterns_*
+directory.
+
```
@QiaolingTang @richm 
Does anyone have a  list of index pattern file names?

mburke5678 commented on install_config/aggregate_logging.adoc:
```diff
@@ -2054,6 +2054,186 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
+    --key /etc/elasticsearch/secret/admin-key \
+    https://localhost:9200/_template/<json-file-name> -XPUT -d@- | \
+python -mjson.tool
+----
```
@QiaolingTang @richm 
What would this command look like to add a JSON to a pod if using project index

mburke5678 commented on install_config/aggregate_logging.adoc:
```diff
@@ -2054,6 +2054,186 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
+----
++
+The index patterns are download to the *_/usr/share/elasticsearch/index_patterns_* directory.
+
```
Would the user need to do this for the OPS cluster or would the index patterns be the same?

mburke5678 commented on install_config/aggregate_logging.adoc:
```diff
@@ -2054,6 +2054,186 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
+----
+"title": "project.MYNAMESPACE.*"
+----
+
```
@QiaolingTang @richm 
For operations, the user would put just `"title": "operations.*"`?

mburke5678 commented:
@QiaolingTang @richm I added some questions in-line.  PTAL.

@QiaolingTang If you have a 3.11 cluster with logging installed that I could look at, I might be able to answer some of my questions.

mburke5678 commented:
@richm Can you help me with this? I am trying to add your blogs on adding custom fields to Kibana to the user docs, at customer request. I don't have access to a 3.11 cluster with logging. I tried the steps in 4.4, but it doesn't work (Different Elastic version?)

https://richm.github.io/add-fields-to-index-template
https://richm.github.io/how-to-add-fields-to-kibana

richm commented:
> @richm Can you help me with this? I am trying to add your blogs on adding custom fields to Kibana to the user docs, at customer request. I don't have access to a 3.11 cluster with logging.

I have a 3.11 system you can use.

> I tried the steps in 4.4, but it doesn't work (Different Elastic version?)

What exactly is not working?

> 
> https://richm.github.io/add-fields-to-index-template
> https://richm.github.io/how-to-add-fields-to-kibana

mburke5678 commented:
> What exactly is not working?
I created the JSON file. 
Added to ES:
```
cat $file | oc exec -n openshift-logging -i -c elasticsearch $espod -- curl -s -k --cert /etc/elasticsearch/secret/admin-cert --key /etc/elasticsearch/secret/admin-key https://localhost:9200/_template/$file -XPUT -d@- | python -mjson.tool
{
    "acknowledged": true
}
```
Ran the command to verify:

```
oc exec -n openshift-logging -i -c elasticsearch $espod -- curl -s -k --cert /etc/elasticsearch/secret/admin-cert --key /etc/elasticsearch/secret/admin-key https://localhost:9200/testy2.*/_search?sort=mytimefield2:desc | python -mjson.tool
{
    "_shards": {
        "failed": 0,
        "skipped": 0,
        "successful": 0,
        "total": 0
    },
    "hits": {
        "hits": [],
        "max_score": 0.0,
        "total": 0
    },
    "timed_out": false,
    "took": 0
} 
```
(Don't like the looks of all those zeros)
Added to Kibana, but am not seeing the custom field mytimefield2

richm commented:
couple of points:

> NOTE The settings will not apply to existing indices. You will need to perform a reindexing for that to work. However, it is usually not a problem, as the settings will apply to new indices

New indices are created at midnight or so, by default.  I suppose one way you could do it would be to create a new namespace, and create a new pod in that namespace, and have something running in that pod that produces JSON output with your new field in the JSON, then configure `MERGE_JSON_LOG=true` so that fluentd will output the field.

> To see if this is working, wait until new indices are created,

See above.

QiaolingTang commented on install_config/aggregate_logging.adoc:
```diff
@@ -2054,6 +2054,244 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
+[[kibana-visualizations-custom]]
+*Adding Custom Fields to Kibana Visualize*
+
+If your {product-title} cluster is configured with custom logging fields
```
What about `If your {product-title} cluster generates logs in json format, which consist of custom fields` ?

QiaolingTang commented on install_config/aggregate_logging.adoc:
```diff
@@ -2054,6 +2054,244 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
+index, you cannot create visualizations with these fields. 
+
+However, if you if you use the `MERGE_JSON_LOG` feature (`MERGE_JSON_LOG=true`) 
+to generate JSON format logs, you can use these custom fields in 
```
I think we don't need to mention `MERGE_JSON_LOG` here as this value is set to `true` by default.

QiaolingTang commented on install_config/aggregate_logging.adoc:
```diff
@@ -2054,6 +2054,244 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
+
+To add custom fields to Kibana Visualize:
+
+. Add custom fields to an Elasticsearch index template:
```
We may need to add a note here: these fields will only be applied to the indices created after the template is updated.

mburke5678 commented:
@QiaolingTang I made the changes as you suggested. PTAL

QiaolingTang commented:
lgtm

mburke5678 commented:
@QiaolingTang Thank you for the quick review!

lamek commented on install_config/aggregate_logging.adoc:
```diff
@@ -2054,6 +2054,252 @@ $ oc exec <$espod> -- es_load_kibana_ui_objects <user-name>
+*Adding Custom Fields to Kibana Visualize*
+
+If your {product-title} cluster generates logs in JSON format that contain custom fields
+that are not are not defined in the Elasticsearch `.operations.\*` or the `project.*` 
```
Extra words here, "are not **are not**"
